src/report/azure-devops.mjs

```diff
@@ -0,0 +1,222 @@
+import { EOL } from "node:os";
+import {
+  formatPercentage,
+  formatViolation as _formatViolation,
+} from "./utl/index.mjs";
+
+const SEVERITY2VSO_TYPE = new Map([
+  // "error" | "warn" | "info" | "ignore"
+  ["error", "error"],
+  ["warn", "warning"],
+  // azure devops doesn't seem to understand 'info'. We still want to
+  // show them, though, hence:
+  ["info", "warning"],
+]);
+
+/**
+ * @param {import("../../types/shared-types.js").SeverityType} pSeverity
+ * @returns {string}
+ */
+function formatSeverity(pSeverity) {
+  return SEVERITY2VSO_TYPE.get(pSeverity) ?? "warning";
+}
+
+/**
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatModuleViolation(pViolation) {
+  return pViolation.from;
+}
+
+/**
+ *
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatDependencyViolation(pViolation) {
+  return `${pViolation.from} -> ${pViolation.to}`;
+}
+
+/**
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatCycleViolation(pViolation) {
+  return `${pViolation.from} -> ${pViolation.cycle.join(" -> ")}`;
+}
+
+/**
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatReachabilityViolation(pViolation) {
+  return `${pViolation.from} -> ${pViolation.to} (via ${pViolation.via.join(
+    " -> "
+  )})`;
+}
+
+/**
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatInstabilityViolation(pViolation) {
+  return `${pViolation.from} -> ${
+    pViolation.to
+  } (instability: ${formatPercentage(
+    pViolation.metrics.from.instability
+  )} -> ${formatPercentage(pViolation.metrics.to.instability)})`;
+}
+
+/**
+ * @param {import("../../types/violations.js").IViolation} pViolation
+ * @returns {string}
+ */
+function formatViolation(pViolation) {
+  const lViolationType2Formatter = {
+    module: formatModuleViolation,
+    dependency: formatDependencyViolation,
+    cycle: formatCycleViolation,
+    reachability: formatReachabilityViolation,
+    instability: formatInstabilityViolation,
+  };
+  let lFormattedViolators = _formatViolation(
+    pViolation,
+    lViolationType2Formatter,
+    formatDependencyViolation
+  );
+  return `##vso[task.logissue type=${formatSeverity(
+    pViolation.rule.severity
+  )};sourcepath=${pViolation.from};code=${
+    pViolation.rule.name
+  };]${lFormattedViolators}${EOL}`;
+}
+
+/**
+ *
+ * @param {number} pNumberOfErrors
+ * @param {number} pNumberOfWarns
+ * @param {number} pNumberOfInfos
+ *
+ * @returns
+ */
+function formatResultStatus(
+  pNumberOfErrors,
+  pNumberOfWarns,
+  pNumberOfInfos,
+  pNumberOfIgnored
+) {
+  if (pNumberOfErrors > 0) {
+    return "Failed";
+  }
+  if (pNumberOfWarns + pNumberOfInfos + pNumberOfIgnored > 0) {
+    return "SucceededWithIssues";
```

Comment:
On second thought, I'm no longer sure we should be using this status here.
Reasoning:
* it will change the status of the entire build even if there's just one warning
* the user has no way of controlling that

I wonder if it wouldn't be "nicer" to let the users decide what to do in some way.
One idea that comes to my mind is to not use this status here and for user to use the [BuildQualityChecks](https://marketplace.visualstudio.com/items?itemName=mspremier.BuildQualityChecks) extension to implement custom quality rules e.g. "set pipeline status to 'SucceededWithIssues' when warnings are > 20" and "fail pipeline when warnings are > 50". This might be more flexible and adaptable to various peoples' needs.

Reply:
Added in 47e143

> It might be nice to add a --max-warnings & --max-ignored to dependency-cruiser, just like [eslint's --max-warnings](https://eslint.org/docs/latest/use/command-line-interface#--max-warnings) 🤔 . Future feature, though.